services/src/datasets/external/gbif.rs

```diff
@@ -197,6 +300,217 @@ impl GbifDataProvider {
+                    union = if index > 0 {
+                        "UNION".to_string()
+                    } else {
+                        String::new()
+                    }
+                )
+            })
+            .reduce(|a, b| a + &b)
+            .unwrap_or_default();
+        let conn = self.pool.get().await?;
+        let query = &format!(
+            r#"
+            WITH names AS (
+                SELECT DISTINCT canonicalname AS name
+                FROM {schema}.species
+                WHERE {taxonrank_filter}{filter} AND canonicalname ILIKE $3
+            )
+            SELECT count, name, rank
+            FROM (
+                {taxonrank_queries}
+            ) AS T
+            ORDER BY name, rank
+            LIMIT $1
+            OFFSET $2;
+            "#,
+            schema = self.db_config.schema,
+            filter = filters
+                .iter()
+                .enumerate()
+                .map(|(index, (column, _))| format!(
+                    r#" AND "{column}" = ${index}"#,
+                    index = index + 4
+                ))
+                .collect::<String>()
+        );
+
+        let stmt = conn.prepare(query).await?;
+
+        let limit = &i64::from(*limit);
+        let offset = &i64::from(*offset);
+        let mut params: Vec<&(dyn ToSql + Sync)> = vec![limit, offset, &search_string];
+        filters.iter().for_each(|(_, value)| params.push(value));
+        let rows = conn.query(&stmt, params.as_slice()).await?;
+
+        Ok(rows
+            .into_iter()
+            .map(|row| {
+                let num_points = row.get::<usize, i64>(0);
+                let name = row.get::<usize, String>(1);
+                let rank = row.get::<usize, String>(2);
+
+                CollectionItem::Layer(LayerListing {
+                    id: ProviderLayerId {
+                        provider_id: GBIF_PROVIDER_ID,
+                        layer_id: LayerId(rank.clone() + "/" + name.as_str()),
+                    },
+                    name: name.clone(),
+                    description: format!(
+                        "{rank}, {num_points} occurrences",
+                        rank = rank[..1].to_ascii_uppercase().to_string() + &rank[1..]
+                    ),
+                    properties: vec![],
+                })
+            })
+            .collect::<Vec<_>>())
+    }
+
+    async fn get_all_datasets_autocomplete_items(
+        &self,
+        limit: &u32,
+        offset: &u32,
+        path: &str,
+        search_string: &str,
+        taxonranks: Vec<String>,
+    ) -> Result<Vec<String>> {
+        let path = path.split_once('/').map_or_else(|| "", |(_, path)| path);
+        let filters = GbifDataProvider::get_filters(path);
+        let taxonrank_filter = Self::create_taxonrank_filter(&taxonranks);
+        let conn = self.pool.get().await?;
+        let query = &format!(
+            r#"
+            SELECT DISTINCT canonicalname
```

Comment:
why is the distinct needed here?

Reply:
There can be duplicates otherwise, for autocomplete they are unwanted

Reply:
Apparently canonicalnames are potentially used for different "taxonomic paths" so there can be duplicates when you don't apply filters along the whole path leading to the taxonrank you're listing